### PR TITLE
ENH: extend factorial{,2,k} to allow complex inputs

### LINKED ISSUE #18409: ENH: extensions of `factorial{,2,k}` for complex domains, recurrences, ...
Following the clean-up of #15600 in #15841, we can now discuss the extensions separately.

## Design Space

There are several dimensions that come into play:
- extensions to negative integers
- extensions to floats
- extensions to complex numbers
- extensions to poles through recurrence (possible e.g. for `factorial2`)

## Extension Issues

Some people would like to have the float/complex extensions of these functions (e..g. https://github.com/scipy/scipy/pull/15349). I tend to agree, but that raises another problem:

#### Complex extensions are incompatible with current inputs

First off, everything that's 0 currently on the negative axis gets a different value. More importantly, all even numbers get smaller by a factor ~0.79=sqrt(2/pi) in the [extension](https://en.wikipedia.org/wiki/Double_factorial#Complex_arguments).

For non-complex floats, there doesn't seem to be another relevant extension, especially not one that matches on all the integers and isn't just a gratuitous hack (this is what #15841 ended up stalling on for a long while). Compared to #15600 I now propose that floats continue to be treated as errors for `extend='zero'` in `factorial{2,k}`.

#### Extensions based on recurrence are incompatible with current status as well as complex extension

This came up for example in #15299. It's [possible](https://en.wikipedia.org/wiki/Double_factorial#Negative_arguments) (for example) to invert the recurrence relation n!! = n * (n-2)!!, and so extend the double-factorial to negative integers. The resulting value would be neither 0 (currently) nor the pole in the complex extension.

## Design Considerations

- Complex extensions  (together with non-positive floats) should be behind a keyword, roughly because:
  - 0-for-negative-values is a valid approach, and has been in scipy ~forever.
  - Since the values being produced are incompatible (for negative & even integers), the complex extension should be explicitly opt-in.
- Using recurrence relations for extending into poles is an open question; not sure if worth it, but if so, probably with another enum value for a potential `extend` keyword.
- Use the most general approximation, except when alternatives with better accuracy / performance exist.

## Possible Future Behaviour (note `extend=`, not `exact=`)

| Inputs | `factorial(`<br/>`extend=`<br/>`'zero')` | `factorial(`<br/>`extend=`<br/>`'complex')` | `factorial2(`<br/>`extend=`<br/>`'zero')` | `factorial2(`<br/>`extend=`<br/>`'complex')` | `factorialk(`<br/>`extend=`<br/>`'zero')` | `factorialk(`<br/>`extend=`<br/>`'complex')` |
|-|-|-|-|-|-|-|
| Scalar positive integer&nbsp;w/ `exact=True` | n! | n! | n!! | n!! | n!(k) | n!(k) |
| Scalar positive integer | Γ(n+1) | Γ(n+1) | [1a] for odd,<br/>[1b] for even<br/>(matches n!!) | [1a] (_doesn't_<br/>match n!!<br/>for even n) | n!(k)<br/>(no approx.) | [2] |
| Scalar negative integer / poles | 0.0 | NaN | 0.0 | NaN | 0.0 | NaN |
| Scalar positive float | Γ(n+1) | Γ(n+1) | `ValueError` | [1a] | `ValueError` | [2] |
| Scalar negative float | 0.0 | Γ(n+1) | `ValueError` | [1a] | `ValueError` | [2] |
| Scalar complex | `ValueError` | Γ(n+1) | `ValueError` | [1a] | `ValueError` | [2] |
| Scalar NaN | NaN | NaN | NaN | NaN | NaN | NaN |
| Positive integer in array | Γ(n+1) | Γ(n+1) |[1a] for odd,<br/>[1b] for even<br/>(matches n!!) | [1a] (_doesn't_<br/>match n!!<br/>for even n) | n!(k)<br/>(no approx.) | [2] |
| Negative integer in array | 0.0 | NaN | 0.0 | NaN | 0.0 | NaN |
| Positive float in array | Γ(n+1) | Γ(n+1) | `ValueError` | [1a] | `ValueError` | [2] |
| Negative float in array | 0.0 | Γ(n+1) | `ValueError` | [1a] | `ValueError` | [2] |
| Complex in array | `ValueError` | Γ(n+1) | `ValueError` | [1a] | `ValueError` | [2] |
| NaN in array | NaN | NaN | NaN | NaN | NaN | NaN |
| Empty array | `array([])` | `array([])` | `array([])` | `array([])` | `array([])` | `array([])` |

The formulas in question are:
* [1a]: `z!! = 2^((z-1)/2) Γ(z/2 + 1) / Γ(1/2 + 1) == 2^(z/2) Γ(z/2 + 1) sqrt(2 / π)` (holds exactly for odd integers and is the basis for the extension)
* [1b]: `z!! = 2^(z/2) Γ(z/2 + 1)` (holds only for even integers)
* [2]: `z!(α) =  α^((z-1)/α) Γ(z/α + 1) / Γ(1/α + 1)` (extension based on α'th integers)

I'd very much welcome feedback on this - there are some trade-offs to be made (e.g. whether negative floats should also raises without `extend='complex'`, etc.)

### PR DESCRIPTION
**Pull request by h-vetinari**
Main implementation for #18409; builds on #19989 and #21702

Closes #18409

Note that there's a lot of simplification possible once the implementation lands, see #19988

### CODE REVIEW

File: scipy/special/_basic.py
Comment (steppi) [inline scipy/special/_basic.py:2949]: @h-vetinari, at this point, I think we should just fix `gamma` and related functions. Hopefully there aren't actually too many users relying on this bug in the behavior at the poles, but if so, it should be fine to just add a note in the docstring explaining the old and new behavior, and to use `rgamma` instead of dividing by the `gamma` function.

I'll make an issue and discourse post and see if we can get the fix through this week.

Sorry, I was the one dragging my feet on fixing this for fear of breaking someones workflow, but seeing the workarounds you had to do has convinced me.

I'll make an issue about it and see if we can get this through

Comment (h-vetinari) [reply scipy/special/_basic.py:2949]: I had completely forgotten about the discussion in #19071 already - that would probably be a good candidate for the issue you're talking about? Happy to switch to `rgamma` of course.

Comment (h-vetinari) [reply scipy/special/_basic.py]: Fixed as discussed in #21827

**Review by steppi — COMMENTED** [review]
Thanks @h-vetinari. This looks really good. No changes requested implementation wise. A couple comments and questions regarding docstrings and warning messages.

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py:3278]: This is a leftover from reshuffling (I've already fixed it locally but didn't want to repush again for a single word change), it should say "below". The factor in question can be seen towards the bottom of the docstring

Comment (steppi) [reply scipy/special/_basic.py:3278]: I see. That makes sense.  Just do a `[docs only]` commit with any docstring changes needed, and then I think this is ready to merge.

Comment (h-vetinari) [reply scipy/special/_basic.py:3278]: > Just do a `[docs only]` commit with any docstring changes needed, and then I think this is ready to merge.

I'd actually prefer to keep the commit history here. I'll push a rebase with the typos you found - thank you for the review!

Comment (steppi) [reply scipy/special/_basic.py:3278]: Sure, no problem. Commit history looks clean.

**Comment by h-vetinari** [comment]
I rebased on main, the failures must be coming from there... I don't think that needs to hold up this PR? If you agree, I'll let you approve and merge @steppi 🙏

**Comment by steppi** [comment]
> I rebased on main, the failures must be coming from there... I don't think that needs to hold up this PR? If you agree, I'll let you approve and merge @steppi 🙏

Yes, failures look unrelated. I'll merge if CI settles with no related failures. Thanks

**Review by steppi — APPROVED** [review]
lgtm

**Pull request merged**
